### ISSUE TITLE
strictFunctionTypes option should strictly check callback parameters of override methods

### ISSUE THREAD

**Issue by falsandtru** [Microsoft/TypeScript#19007]
As @aluanhaddad mentioned at https://github.com/Microsoft/TypeScript/issues/19004#issuecomment-334908744.

cc @ahejlsberg 

**TypeScript Version:**  master

**Code**

```ts
class C {
  a(f: (a: C) => C): void {
  }
}
class D extends C {
  p: boolean = true;
  a(f: (a: D) => D): void { // should throw an error here.
    console.log(this.p);
    this.p = f(new D()).p; // type is boolean, but value is undefined.
    console.log(this.p);
  }
}

const c: C = new D();
c.a(() => new C());
```

**Expected behavior:**

error

**Actual behavior:**

```
$ node built/local/tsc.js index.ts --strictFunctionTypes && node index.js
true
undefined
```

**Comment by falsandtru** [comment]
cc @ahejlsberg

**Comment by ahejlsberg** [comment]
This is working as intended. As explained [here](https://github.com/Microsoft/TypeScript/pull/18654#issue-259565697), in `--strictFunctionTypes` mode we still check parameters of methods and constructors bivariantly. So, there is no error here:

```ts
interface C {
  a(f: (a: C) => C): void;
}
interface D extends C {
  p: boolean;
  a(f: (a: D) => D): void;
}
```

But an error is reported here:

```ts
interface C {
  a: (f: (a: C) => C) => void;
}
interface D extends C {
  p: boolean;
  a: (f: (a: D) => D) => void;
}
```

As I explain [here](https://github.com/Microsoft/TypeScript/issues/18963#issuecomment-334586832), we can't tighten this any further without causing `Array<T>` and other core types to become invariant.

**Comment by falsandtru** [comment]
I see, thanks for your explanation.
